### PR TITLE
Fix for null item crash introduced in PR #857

### PR DESCRIPTION
**Pull request by jarmengaud**
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

Regression of PR #857:
We cannot unequip a weapon (main / sub and range slots) without crashing the server

### CODE REVIEW

File: src/map/utils/charutils.cpp [hunk 1]
```diff
@@ -2186,8 +2186,15 @@ namespace charutils
         }
         if (equipSlotID == SLOT_MAIN || equipSlotID == SLOT_RANGED || equipSlotID == SLOT_SUB)
         {
-            // If the weapon ISN'T an wind based instrument or an string based instrument
-            if (((CItemWeapon*)PItem)->getSkillType() != SKILL_STRING_INSTRUMENT && ((CItemWeapon*)PItem)->getSkillType() != SKILL_WIND_INSTRUMENT)
+            if ((PItem != nullptr) && PItem->isType(ITEM_EQUIPMENT))
+            {
+                // If the weapon ISN'T an wind based instrument or an string based instrument
+                if (((CItemWeapon*)PItem)->getSkillType() != SKILL_STRING_INSTRUMENT && ((CItemWeapon*)PItem)->getSkillType() != SKILL_WIND_INSTRUMENT)
+                {
```
Comment (Kreidos) [inline src/map/utils/charutils.cpp]: This if just contains another if... Which does the same as the else statement below... I think we can simplify this to just one if with no else statement at all.

File: src/map/utils/charutils.cpp [hunk 2]
```diff
@@ -2186,8 +2186,15 @@ namespace charutils
         }
         if (equipSlotID == SLOT_MAIN || equipSlotID == SLOT_RANGED || equipSlotID == SLOT_SUB)
         {
-            // If the weapon ISN'T an wind based instrument or an string based instrument
-            if (((CItemWeapon*)PItem)->getSkillType() != SKILL_STRING_INSTRUMENT && ((CItemWeapon*)PItem)->getSkillType() != SKILL_WIND_INSTRUMENT)
+            if ((PItem != nullptr) && PItem->isType(ITEM_EQUIPMENT))
```
Comment (TeoTwawki) [inline src/map/utils/charutils.cpp]: `if ((PItem)` is the same as `if ((PItem != nullptr)` and I think has been the format the project has been favoring (someone correct me if I am wrong)

**Review by Kreidos — CHANGES_REQUESTED** [review]
Just a little restructuring would be nice to make the intention here a bit clearer. :)

File: src/map/utils/charutils.cpp [hunk 3]
```diff
@@ -2186,8 +2186,15 @@ namespace charutils
...
+            {
+                // If the weapon ISN'T an wind based instrument or an string based instrument
+                if (((CItemWeapon*)PItem)->getSkillType() != SKILL_STRING_INSTRUMENT && ((CItemWeapon*)PItem)->getSkillType() != SKILL_WIND_INSTRUMENT)
+                {
+                    PChar->health.tp = 0;
+                }
+            }
+            else
```
Comment (TeoTwawki) [inline src/map/utils/charutils.cpp:2197]: Not sure under what condition you would hit this else block that would be retail accurate tp loss.

**Review by Kreidos — APPROVED** [review]

File: src/map/utils/charutils.cpp [hunk 4]
```diff
@@ -2186,9 +2186,11 @@ namespace charutils
         }
         if (equipSlotID == SLOT_MAIN || equipSlotID == SLOT_RANGED || equipSlotID == SLOT_SUB)
         {
-            // If the weapon ISN'T an wind based instrument or an string based instrument
-            if (((CItemWeapon*)PItem)->getSkillType() != SKILL_STRING_INSTRUMENT && ((CItemWeapon*)PItem)->getSkillType() != SKILL_WIND_INSTRUMENT)
+            if (!PItem || !PItem->isType(ITEM_EQUIPMENT) ||
```
Comment (TeoTwawki) [inline src/map/utils/charutils.cpp:2189]:
```c++
auto PWeapon = dynamic_cast<CItemWeapon*>(PItem)
if (PWeapon && PWeapon->getSkillType() != SKILL_STRING_INSTRUMENT && PWeapon->getSkillType() != SKILL_WIND_INSTRUMENT)
```
That's what was meant on discord about casting. I haven't tested that and it was suggested to me by a 3rd party

Comment (Kreidos) [reply src/map/utils/charutils.cpp:2189]: This would work too; minor change on such a small fix but a bit cleaner.

Comment (zach2good) [reply src/map/utils/charutils.cpp:2189]: +1 on this recommendation, a rare and beautiful possible use of `dynamic_cast`

**Review by zach2good — APPROVED** [review]

**Review by 59blargedy — APPROVED** [review]

**Review by Wiggo32 — APPROVED** [review]

**Pull request merged**